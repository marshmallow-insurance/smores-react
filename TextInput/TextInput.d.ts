import React, { FocusEvent, FormEvent } from 'react';
import { CommonFieldProps } from '../fields/commonFieldTypes';
interface Props extends CommonFieldProps {
    type?: 'text' | 'email' | 'password' | 'time' | 'date' | 'tel';
    placeholder: string;
    name?: string;
    value: string;
    onBlur?: (e: FocusEvent<HTMLInputElement>) => void;
}
/** on change or on input required */
type InputProps = {
    /** on change is required and on input optional */
    onChange: (e: string) => void;
    onInputChange?: (e: FormEvent<HTMLInputElement>) => void;
} | {
    /** on input is required and on change optional */
    onChange?: (e: string) => void;
    onInputChange: (e: FormEvent<HTMLInputElement>) => void;
};
export type TextInputProps = Props & InputProps;
export declare const TextInput: React.ForwardRefExoticComponent<TextInputProps & React.RefAttributes<HTMLInputElement>>;
export {};
