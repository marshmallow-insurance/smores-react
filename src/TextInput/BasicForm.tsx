import React, {SFC, useState} from 'react';
import {TextInput} from './TextInput';

type Props = {};

const BasicForm: SFC<Props> = () => {
  const [value, setValue] = useState('');

  return (
    <form>
      <TextInput
        id="textInput"
        label="label"
        name="textInput"
        onChange={(str: string) => setValue(str)}
        placeholder="Placeholder"
        error={value.length > 7}
        errorMsg="Value is over 7 characters!"
        trailingIcon="contact-at"
        value={value}
      />
    </form>
  );
};

export default BasicForm;
