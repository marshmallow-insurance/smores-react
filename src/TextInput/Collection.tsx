import React, { FC, ReactNode, useState } from 'react'
import styled from 'styled-components'

import { Text } from '../Text'
import { Box } from '../Box'
import { TextInput, TextInputProps } from './TextInput'
import { noop } from '../utils/noop'
import { SupportMessage } from '../SupportMessage'

const variationList: TextInputProps[] = [
  {
    id: 'days',
    placeholder: 'Enter name...',
    errorMsg: 'This field is required',
    assistiveText: (
      <>
        This is a description with a{' '}
        <a href="marshmallow.com" rel="noopener noreferrer" target="_blank">
          link
        </a>
        .
      </>
    ),
    label: 'First name',
    onChange: noop,
    onInputChange: noop,
    onBlur: noop,
    value: '',
  },
]

export const CollectionPage: FC = () => {
  const [, setValue] = useState('')

  return (
    <Box flex direction="column">
      {variationList.map(
        ({ id, placeholder, errorMsg, label, assistiveText }) => {
          return (
            <Box key={id} flex direction="column" mb="32px">
              <Title tag="span" typo="header-medium">
                TextInput variations
              </Title>
              <Row label="Generic">
                <TextInput
                  id={id}
                  label={label}
                  value=""
                  placeholder={placeholder}
                  error={false}
                  errorMsg={errorMsg}
                  onChange={setValue}
                />
              </Row>
              <Row label="With label as title">
                <TextInput
                  id={id}
                  label={label}
                  value=""
                  renderAsTitle
                  placeholder={placeholder}
                  error={false}
                  errorMsg={errorMsg}
                  onChange={setValue}
                />
              </Row>
              <Row label="Fallback">
                <TextInput
                  id={id}
                  label={label}
                  value=""
                  placeholder={placeholder}
                  fallbackStyle={true}
                  error={false}
                  errorMsg={errorMsg}
                  onChange={setValue}
                />
              </Row>
              <Row label="Disabled">
                <TextInput
                  id={id}
                  label={label}
                  value=""
                  placeholder={placeholder}
                  disabled={true}
                  error={false}
                  errorMsg={errorMsg}
                  onChange={setValue}
                />
              </Row>
              <Row label="Required">
                <TextInput
                  id={id}
                  label={label}
                  value=""
                  required={true}
                  placeholder={placeholder}
                  error={false}
                  errorMsg={errorMsg}
                  onChange={setValue}
                />
              </Row>
              <Row label="Leading Icon">
                <TextInput
                  id={id}
                  label={label}
                  frontIcon="search"
                  value=""
                  placeholder={placeholder}
                  error={false}
                  errorMsg={errorMsg}
                  onChange={setValue}
                />
              </Row>
              <Row label="Trailing Icon">
                <TextInput
                  id={id}
                  label={label}
                  trailingIcon="car"
                  value=""
                  placeholder={placeholder}
                  error={false}
                  errorMsg={errorMsg}
                  onChange={setValue}
                />
              </Row>
              <Row label="Assistive text">
                <TextInput
                  id={id}
                  label={label}
                  value=""
                  placeholder={placeholder}
                  assistiveText="Some assistive text"
                  error={false}
                  errorMsg={errorMsg}
                  onChange={setValue}
                />
              </Row>
              <Row label="With label as title and assistive text with link">
                <TextInput
                  id={id}
                  label={label}
                  value=""
                  renderAsTitle
                  assistiveText={assistiveText}
                  placeholder={placeholder}
                  error={false}
                  completed={true}
                  errorMsg={errorMsg}
                  onChange={setValue}
                />
              </Row>
              <Row label="Completed">
                <TextInput
                  id={id}
                  label={label}
                  value=""
                  placeholder={placeholder}
                  error={false}
                  completed={true}
                  errorMsg={errorMsg}
                  onChange={setValue}
                />
              </Row>
              <Row label="As title">
                <TextInput
                  id={id}
                  label={label}
                  value=""
                  renderAsTitle={true}
                  placeholder={placeholder}
                  error={false}
                  errorMsg={errorMsg}
                  onChange={setValue}
                />
              </Row>
              <Row label="Error">
                <TextInput
                  id={id}
                  label={label}
                  value=""
                  required={true}
                  placeholder={placeholder}
                  assistiveText="Some assistive text"
                  error={true}
                  errorMsg={errorMsg}
                  onChange={setValue}
                />
              </Row>
              <Row label="React element error">
                <TextInput
                  id={id}
                  label={label}
                  required={true}
                  value=""
                  placeholder={placeholder}
                  error={true}
                  errorMsg={
                    <SupportMessage type="warning" description="error!!" />
                  }
                  onChange={setValue}
                />
              </Row>
              <Row label="Fallback Error">
                <TextInput
                  id={id}
                  label={label}
                  required={true}
                  value=""
                  placeholder={placeholder}
                  fallbackStyle={true}
                  error={true}
                  errorMsg={errorMsg}
                  onChange={setValue}
                />
              </Row>
            </Box>
          )
        },
      )}
    </Box>
  )
}

const Row: FC<{ label: string; children?: ReactNode }> = ({
  label,
  children,
}) => {
  return (
    <RowWrapper mt="32px">
      <Box width="140px">
        <Text tag="span" typo="label-large">
          {label}
        </Text>
      </Box>
      {children}
    </RowWrapper>
  )
}

const RowWrapper = styled(Box)`
  display: flex;
  flex-direction: row;
  gap: 16px;
`

const LABEL_WIDTH = '140px'

const Title = styled(Text)`
  margin-bottom: 12px;
  font-size: 18px;
  line-height: 20px;
  margin-left: ${LABEL_WIDTH};
`
