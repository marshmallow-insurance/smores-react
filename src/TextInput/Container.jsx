import React, {useState} from 'react';
import {TextInput} from './TextInput';

export const Container = () => {
  const [value, setValue] = useState('');

  return (
    <form>
      <TextInput
        id="textInput"
        label="label"
        name="textInput"
        onChange={str => setValue(str)}
        placeholder="Placeholder"
        error={value.length > 7}
        errorMsg="Value is over 7 characters!"
        trailingIcon="contact-at"
        value={value}
      />
    </form>
  );
};
