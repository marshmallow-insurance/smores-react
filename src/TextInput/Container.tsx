import React, { useState } from 'react'
import { TextInput } from './TextInput'

export const Container = () => {
  const [value, setValue] = useState('')

  return (
    <form>
      <TextInput
        id="textInput"
        label="label"
        name="textInput"
        onChange={setValue}
        placeholder="Placeholder"
        error={value.length > 7}
        errorMsg="Value is over 7 characters!"
        trailingIcon="at"
        value={value}
      />
    </form>
  )
}
