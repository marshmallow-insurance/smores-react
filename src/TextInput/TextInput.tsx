import React, { FormEvent, forwardRef, ForwardedRef } from 'react'
import styled from 'styled-components'
import { darken } from 'polished'

import { Box } from '../Box'
import { Text } from '../Text'
import { Icon } from '../Icon'

import { theme } from '../theme'
import { useUniqueId } from '../utils/id'

type DefaultProps = {
  id?: string
  className?: string
  /** Input type for proper browser support */
  type?: 'text' | 'email' | 'password' | 'time' | 'date'
  /** used to render outlined style */
  outlined?: boolean
  /** Placeholder */
  placeholder: string
  /** label displayed above the input  */
  label?: string
  /** used for label - input connection */
  name?: string
  /** input value */
  value: string
  /** error flag */
  error?: boolean
  /** error text message */
  errorMsg?: string
  /** onBlur listener */
  onBlur?: (e: FormEvent<HTMLInputElement>) => void
  /** used for adding a trailing icon */
  trailingIcon?: string
  /** Disabled flag */
  disabled?: boolean
  /** Required flag */
  required?: boolean
}

/** on change or on input required */
type TruncateProps =
  | {
      /** on change is required and on input optional */
      onChange: (e: string) => void
      onInputChange?: (e: FormEvent<HTMLInputElement>) => void
    }
  | {
      /** on input is required and on change optional */
      onChange?: (e: string) => void
      onInputChange: (e: FormEvent<HTMLInputElement>) => void
    }

export type TextInputProps = DefaultProps & TruncateProps

export const TextInput = forwardRef(function TextInput(
  {
    id: idProp,
    className = '',
    type = 'text',
    placeholder,
    label,
    name,
    value,
    outlined = false,
    error = false,
    errorMsg,
    trailingIcon,
    onBlur,
    onChange,
    onInputChange,
    disabled = false,
    required = false,
  }: TextInputProps,
  ref: ForwardedRef<HTMLInputElement>,
) {
  const id = useUniqueId(idProp)

  return (
    <Container className={className} hasLabel={!!label} hasError={!!errorMsg}>
      {label && (
        <Box mb={outlined ? '4px' : '0px'}>
          <Text tag="label" color="subtext" typo="label" htmlFor={id}>
            {label}
          </Text>
          {required && (
            <Text tag="label" color="error" typo="label">
              *
            </Text>
          )}
        </Box>
      )}

      <Content value={value} outlined={outlined} error={error}>
        <Input
          disabled={disabled}
          type={type}
          id={id}
          name={name}
          ref={ref}
          placeholder={placeholder}
          value={value}
          error={error}
          outlined={outlined}
          autoComplete="off"
          onChange={(e: FormEvent<HTMLInputElement>) => {
            onChange && onChange(e.currentTarget.value)
            onInputChange && onInputChange(e)
          }}
          onBlur={(e) => {
            onBlur && onBlur(e)
          }}
        />
        {trailingIcon && <Icon render={trailingIcon} color="subtext" />}
      </Content>
      {error && <ErrorBox>{errorMsg}</ErrorBox>}
    </Container>
  )
})

interface IContainer {
  hasLabel: boolean
  hasError: boolean
}

const Container = styled.div<IContainer>`
  display: flex;
  flex-direction: column;
  height: auto;
`

interface IInput {
  error: boolean
  disabled?: boolean
  value?: string
  outlined?: boolean
}

const Content = styled.div<IInput>`
  border-bottom: 1px solid;
  border-color: ${({ error }) =>
    theme.colors[`${error ? 'error' : 'outline'}`]};
  background-color: ${({ outlined }) =>
    outlined ? 'transparent' : theme.colors['white']};
  display: flex;
  height: 32px;

  &:hover,
  &:focus-within {
    border-color: ${({ error }) =>
      error ? theme.colors.error : darken(0.1, theme.colors.outline)};
  }

  ${({ outlined, error }) =>
    outlined &&
    `
      border: 2px solid ${error ? theme.colors.error : theme.colors.outline};
      border-radius: 8px;
      height: auto;
    `}

  ${({ value }) =>
    value &&
    value != '' &&
    `
      border-color: ${theme.colors.outline};
    `}
`

const Input = styled.input<IInput>`
  border: none;
  background-color: transparent;
  color: ${({ error }) => theme.colors[`${error ? 'error' : 'secondary'}`]};
  font-size: 16px;
  width: 100%;
  outline: none;
  cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'initial')};
  opacity: ${({ disabled }) => (disabled ? '0.5' : '1')};
  padding: ${({ outlined }) => (outlined ? '17px 14px' : '1px 2px')};

  &::placeholder {
    color: ${theme.colors.subtext};
  }
`

const ErrorBox = styled.span`
  margin-top: 7px;
  color: ${theme.colors.error};
  font-size: 12px;
`
