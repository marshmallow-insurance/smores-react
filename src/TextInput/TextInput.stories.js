import React from 'react';
import { storiesOf } from '@storybook/react';

import { text, boolean } from '@storybook/addon-knobs';

import TextInput from './TextInput';
import BasicForm from './BasicForm';

storiesOf('TextInput', module)
  .addWithJSX('default', () => (
    <TextInput
      id='textInput'
      name='textInput'
      placeholder={text('Placeholder', 'Placeholder')}
      value=''
      onChange={(str) => {}}
    />
  ))
  .addWithJSX('error', () => (
    <TextInput
      id='textInput'
      name='textInput'
      label={text('Label', 'label')}
      placeholder={text('Placeholder', 'Placeholder')}
      value='kek'
      error={boolean('Error', true)}
      errorMsg='Oh boy, something went wrong!'
      onChange={(str) => {}}
    />
  ))
  .addWithJSX('with Label', () => (
    <TextInput
      id='textInput'
      name='textInput'
      label={text('Label', 'label')}
      placeholder={text('Placeholder', 'Placeholder')}
      value=''
      onChange={(str) => {}}
    />
  ))
  .addWithJSX('with Icon', () => (
    <TextInput
      id='textInput'
      name='textInput'
      label={text('Label', 'label')}
      placeholder={text('Placeholder', 'Placeholder')}
      value='kek'
      trailingIcon='contact-at'
      onChange={(str) => {}}
    />
  ))
  .addWithJSX('functional', () => (
    <BasicForm />
  ))