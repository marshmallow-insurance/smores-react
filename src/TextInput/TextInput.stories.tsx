import React from 'react'
import { TextInput, TextInputProps } from './TextInput'
import { SupportMessage } from '../SupportMessage'
import { Container } from './Container'
import { noop } from '../utils/noop'

export default {
  title: 'Text Input',
  component: TextInput,
  argTypes: { onChange: { action: 'changed' } },
}

const Template = (props: TextInputProps) => <TextInput {...props} />

export const Default = Template.bind({})

Default.args = {
  id: 'textInput',
  name: 'textInput',
  placeholder: 'Placeholder text',
  onChange: noop,
  onInputChange: noop,
  onBlur: noop,
}

export const DefaultFallback = Template.bind({})

DefaultFallback.args = {
  id: 'textInput',
  name: 'textInput',
  placeholder: 'Placeholder text',
  fallback: true,
  onChange: noop,
  onInputChange: noop,
  onBlur: noop,
}

export const TypePassword = Template.bind({})

TypePassword.args = {
  id: 'textInput',
  name: 'textInput',
  placeholder: 'Placeholder text',
  onChange: noop,
  onInputChange: noop,
  onBlur: noop,
  type: 'password',
}

export const Required = Template.bind({})

Required.args = {
  id: 'textInput',
  name: 'textInput',
  placeholder: 'Enter name',
  onChange: noop,
  onInputChange: noop,
  onBlur: noop,
  label: 'Please enter your name',
  required: true,
}

export const WithError = Template.bind({})

WithError.args = {
  id: 'textInput',
  name: 'textInput',
  placeholder: 'Placeholder text',
  onChange: noop,
  onInputChange: noop,
  onBlur: noop,
  required: true,
  error: true,
  errorMsg: 'Oh boy, something went wrong!',
}

export const Disabled = Template.bind({})

Disabled.args = {
  id: 'textInput',
  name: 'textInput',
  placeholder: 'Placeholder text',
  onChange: noop,
  onInputChange: noop,
  onBlur: noop,
  label: 'label',
  disabled: true,
}

export const DisabledFallback = Template.bind({})

DisabledFallback.args = {
  id: 'textInput',
  name: 'textInput',
  placeholder: 'Placeholder text',
  onChange: noop,
  onInputChange: noop,
  onBlur: noop,
  label: 'label',
  fallback: true,
  disabled: true,
}

export const WithLabel = Template.bind({})

WithLabel.args = {
  id: 'textInput',
  name: 'textInput',
  placeholder: 'Placeholder text',
  onChange: noop,
  onInputChange: noop,
  onBlur: noop,
  label: 'label',
}

export const WithFrontIcon = Template.bind({})

WithFrontIcon.args = {
  id: 'textInput',
  name: 'textInput',
  placeholder: 'Placeholder text',
  onChange: noop,
  onInputChange: noop,
  onBlur: noop,
  label: 'label',
  frontIcon: 'at',
}

export const WithTrailingIcon = Template.bind({})

WithTrailingIcon.args = {
  id: 'textInput',
  name: 'textInput',
  placeholder: 'Placeholder text',
  onChange: noop,
  onInputChange: noop,
  onBlur: noop,
  label: 'label',
  trailingIcon: 'at',
}

export const LabelAsTitle = Template.bind({})

LabelAsTitle.args = {
  id: 'textInput',
  name: 'textInput',
  placeholder: 'Enter name',
  onChange: noop,
  onInputChange: noop,
  onBlur: noop,
  label: 'Please enter your name',
  required: true,
  assistiveText: 'Some more information...',
  renderAsTitle: true,
}

export const WithReactElementError = Template.bind({})

WithReactElementError.args = {
  id: 'textInput',
  name: 'textInput',
  label: 'with Support Message as Error',
  placeholder: 'Placeholder text',
  onChange: noop,
  onInputChange: noop,
  onBlur: noop,
  error: true,
  errorMsg: <SupportMessage type="warning" description="error!!" />,
}

const WorkingExampleTemplate = () => <Container />

export const WorkingExample = WorkingExampleTemplate.bind({})
