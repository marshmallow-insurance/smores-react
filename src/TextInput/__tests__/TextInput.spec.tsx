import { render } from '@testing-library/react'
import React from 'react'

import { TextInput } from '../TextInput'

test('renders', () => {
  const { container } = render(
    <TextInput
      id="textInput"
      name="textInput"
      placeholder="Placeholder"
      value=""
      onChange={(str) => console.log(str)}
    />,
  )
  expect(container.firstChild).toMatchSnapshot()
})
