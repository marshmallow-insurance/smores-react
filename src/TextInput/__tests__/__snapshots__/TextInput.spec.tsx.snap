// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders 1`] = `
<div
  class="sc-aXZVg sc-gsFSXq byxNQM cZVmFI"
>
  <div
    class="sc-aXZVg byxNQM"
  >
    <div
      class="sc-aXZVg cowLnB"
    >
      <input
        autocomplete="off"
        class="sc-kAyceB jUthIA"
        id="textInput"
        name="textInput"
        placeholder="Placeholder"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
